### PR TITLE
Fix permissions grid scrollbar issues

### LINKED ISSUE #3512: Missing horizontal scrollbar, and phantom unnecessary scrollbar 👻
On both Chrome and Safari on release-0.20, in the new permissions table, if I have lots of groups there's no scrollbar visible to allow me to scroll to the right to see them (though I can scroll using the laptop touchpad). The expected behavior would be an always-visible horizontal scrollbar somewhere on the screen in the event it's needed.

I'm also seeing a weird phantom vertical scrollbar that shouldn't be there:

lots-of-groups

### PR DESCRIPTION
**Pull request by tlrobinson**
Resolves #3512

### CODE REVIEW

**Comment by tlrobinson** [comment]
@metabase/core-developers Review plz

**Comment by kdoh** [comment]
Seems to make sense. I'm a bit itchy about the whole phantom cell business but overall I think this is 👍

**Comment by tlrobinson** [comment]
@bvaughn is there a better solution to adding a bit of padding in a grid than this hack:

```
<Grid
  rowCount={rowCount + 1}
  rowHeight={({ index }) => index >= rowCount ? paddingBottom : rowHeight}
  cellRenderer={({ key, style, columnIndex, rowIndex }) =>
    rowIndex >= rowCount ? null : cellRenderer({ key, style, columnIndex, rowIndex })
  }
  ...
/>
```

**Comment by bvaughn** [comment]
Unless I'm misunderstanding your intentions, yes. I think there's a better way. Why not...

```diff
39c39
<                             width={width - rowHeaderWidth}
---
>                             width={width - rowHeaderWidth - paddingRight}
43,44c43
<                                     {/*  HACK: pad the right with a phantom cell */}
<                                     {columnIndex >= columnCount ? null : renderColumnHeader({ columnIndex })}
---
>                                     {renderColumnHeader({ columnIndex })}}
47,48c46,47
<                             columnCount={columnCount + 1}
<                             columnWidth={({ index }) => index >= columnCount ? paddingRight : columnWidth}
---
>                             columnCount={columnCount}
>                             columnWidth={columnWidth}
59c58
<                             height={height - columnHeaderHeight}
---
>                             height={height - columnHeaderHeight - paddingBottom}
62,63c61
<                                     {/*  HACK: pad the bottom with a phantom cell */}
<                                     {rowIndex >= rowCount ? null : renderRowHeader({ rowIndex })}
---
>                                     {renderRowHeader({ rowIndex })}
68,69c66,67
<                             rowCount={rowCount + 1}
<                             rowHeight={({ index }) => index >= rowCount ? paddingBottom : rowHeight}
---
>                             rowCount={rowCount}
>                             rowHeight={rowHeight}
77,78c75,76
<                             width={width - rowHeaderWidth}
<                             height={height - columnHeaderHeight}
---
>                             width={width - rowHeaderWidth - paddingRight}
>                             height={height - columnHeaderHeight - paddingBottom}
81,82c79
<                                     {/*  HACK: pad the bottom/right with a phantom cell */}
<                                     {rowIndex >= rowCount || columnIndex >= columnCount ? null : renderCell({ columnIndex, rowIndex })}
---
>                                     {renderCell({ columnIndex, rowIndex })}
85,88c82,85
<                             columnCount={columnCount + 1}
<                             columnWidth={({ index }) => index >= columnCount ? paddingRight : columnWidth}
<                             rowCount={rowCount + 1}
<                             rowHeight={({ index }) => index >= rowCount ? paddingBottom : rowHeight}
---
>                             columnCount={columnCount}
>                             columnWidth={columnWidth}
>                             rowCount={rowCount}
>                             rowHeight={rowHeight}
```

**Comment by bvaughn** [comment]
Essentially the easiest/best way to pad a `Grid` is to pad its parent (if you're using `AutoSizer`) _or_ to just manually subtract the padding (if you're manually managing sizes like in this PR).

**Comment by tlrobinson** [comment]
@bvaughn Thanks for your quick response.

I actually want the padding to be inside the scroll container, so you only see the padding once you've scrolled to the end, like so:

Scrolled to top left:

screenshot 2016-10-12 16 31 55

Scrolled to bottom right:

screenshot 2016-10-12 16 32 28

Your solution pads the entire grid like this:

Scrolled to top left:

screenshot 2016-10-12 16 32 07

Scrolled to bottom right:

screenshot 2016-10-12 16 32 14

**Comment by bvaughn** [comment]
Ah. I see!

In that case, I think the hack you've proposed may be reasonable. You could
also put the padding inside the last column/row but that's not an obvious
improvement.

I have not yet encountered this use case. :smile: Let me give it some more
thought on my run. If anything better occurs to me I'll come back.

File: frontend/src/metabase/admin/permissions/components/FixedHeaderGrid.jsx [hunk 1]
```diff
@@ -85,14 +78,14 @@ const FixedHeaderGrid = ({
...
                                 </div>
                             }
-                            columnCount={columnCount}
+                            columnCount={columnCount + 1}
+                            columnWidth={({ index }) => index >= columnCount ? paddingRight : columnWidth}
                             rowCount={rowCount + 1}
-                            columnWidth={columnWidth}
-                            rowHeight={(index) => index >= rowCount ? paddingBottom : rowHeight}
+                            rowHeight={({ index }) => index >= rowCount ? paddingBottom : rowHeight}
                             onScroll={({ scrollTop, scrollLeft }) => onScroll({ scrollTop, scrollLeft })}
```
Comment (bvaughn) [inline frontend/src/metabase/admin/permissions/components/FixedHeaderGrid.jsx]: nit: These `onScroll` wrappers seem unnecessary.

**Comment by bvaughn** [comment]
I believe I have a better proposal, although it would depend on a minor update to react-virtualized. I would be happy to add the feature though, and push out a minor release tonight. Let me know what you think. Here's the idea: `Grid` gets a new optional prop, `containerStyle`. Your code then changes like so:

```jsx
<Grid
  width={width - rowHeaderWidth}
  height={height - columnHeaderHeight}
  cellRenderer={({ key, style, columnIndex, rowIndex }) =>
    <div key={key} style={style}>
      {renderCell({ columnIndex, rowIndex })}
    </div>
  }
  columnCount={columnCount}
  columnWidth={columnWidth}
  containerStyle={{
    width: columnCount * columnWidth + paddingRight,
    height: rowCount * rowHeight + paddingBottom,
    maxWidth: columnCount * columnWidth + paddingRight,
    maxHeight: rowCount * rowHeight + paddingBottom
  }}
  rowCount={rowCount}
  rowHeight={rowHeight}
  onScroll={onScroll}
  scrollTop={scrollTop}
  scrollLeft={scrollLeft}
/>
```

It could even be further simplified to:

```jsx
<Grid
  width={width - rowHeaderWidth}
  height={height - columnHeaderHeight}
  cellRenderer={({ key, style, columnIndex, rowIndex }) =>
    <div key={key} style={style}>
      {renderCell({ columnIndex, rowIndex })}
    </div>
  }
  {...rest}
/>
```

**Comment by bvaughn** [comment]
Had hoped to wait for feedback from your side before this was released but I'm up against a bit of a time crunch because I'll soon be losing my dev laptop for a couple of days (in between jobs).

I've gone ahead and added the `containerStyle` prop to `Grid` in version 8.1.0. I hope this helps simplify your integration. Cheers~

**Comment by tlrobinson** [comment]
@bvaughn Thanks, containerStyle seems to mostly work, except when you scroll all the way to the end ScrollSync seems to stop short, resulting in misalignment of the two grids. I actually ran into the same issue when I was working on the hack currently in this PR, but resolved it at some point (not sure how)
 
screenshot 2016-10-13 15 23 06

**Comment by bvaughn** [comment]
You need to apply a `containerStyle` to your fixed header row / left column as well. 😄

**Comment by bvaughn** [comment]
I only provided an example for the primary `Grid` ~ since the others would be subsets.

**Comment by tlrobinson** [comment]
@bvaughn here's my diff

```
diff --git a/frontend/src/metabase/admin/permissions/components/FixedHeaderGrid.jsx b/frontend/src/metabase/admin/permissions/components/FixedHeaderGrid.jsx
index 4de3c68..a492437 100644
--- a/frontend/src/metabase/admin/permissions/components/FixedHeaderGrid.jsx
+++ b/frontend/src/metabase/admin/permissions/components/FixedHeaderGrid.jsx
@@ -40,14 +40,17 @@ const FixedHeaderGrid = ({
                             height={columnHeaderHeight}
                             cellRenderer={({ key, style, columnIndex, rowIndex }) =>
                                 <div key={key} style={style}>
-                                    {/*  HACK: pad the right with a phantom cell */}
-                                    {columnIndex >= columnCount ? null : renderColumnHeader({ columnIndex })}
+                                    {renderColumnHeader({ columnIndex })}
                                 </div>
                             }
-                            columnCount={columnCount + 1}
-                            columnWidth={({ index }) => index >= columnCount ? paddingRight : columnWidth}
+                            columnCount={columnCount}
+                            columnWidth={columnWidth}
                             rowCount={1}
                             rowHeight={columnHeaderHeight}
+                            containerStyle={{
+                                width: columnCount * columnWidth + paddingRight,
+                                maxWidth: columnCount * columnWidth + paddingRight,
+                            }}
                             onScroll={({ scrollLeft }) => onScroll({ scrollLeft })}
                             scrollLeft={scrollLeft}
                         />
@@ -59,14 +62,17 @@ const FixedHeaderGrid = ({
                             height={height - columnHeaderHeight}
                             cellRenderer={({ key, style, columnIndex, rowIndex }) =>
                                 <div key={key} style={style}>
-                                    {/*  HACK: pad the bottom with a phantom cell */}
-                                    {rowIndex >= rowCount ? null : renderRowHeader({ rowIndex })}
+                                    {renderRowHeader({ rowIndex })}
                                 </div>
                             }
                             columnCount={1}
                             columnWidth={rowHeaderWidth}
-                            rowCount={rowCount + 1}
-                            rowHeight={({ index }) => index >= rowCount ? paddingBottom : rowHeight}
+                            rowCount={rowCount}
+                            rowHeight={rowHeight}
+                            containerStyle={{
+                                height: rowCount * rowHeight + paddingBottom,
+                                maxHeight: rowCount * rowHeight + paddingBottom
+                            }}
                             onScroll={({ scrollTop }) => onScroll({ scrollTop })}
                             scrollTop={scrollTop}
                         />
@@ -78,14 +84,19 @@ const FixedHeaderGrid = ({
                             height={height - columnHeaderHeight}
                             cellRenderer={({ key, style, columnIndex, rowIndex }) =>
                                 <div key={key} style={style}>
-                                    {/*  HACK: pad the bottom/right with a phantom cell */}
-                                    {rowIndex >= rowCount || columnIndex >= columnCount ? null : renderCell({ columnIndex, rowIndex })}
+                                    {renderCell({ columnIndex, rowIndex })}
                                 </div>
                             }
-                            columnCount={columnCount + 1}
-                            columnWidth={({ index }) => index >= columnCount ? paddingRight : columnWidth}
-                            rowCount={rowCount + 1}
-                            rowHeight={({ index }) => index >= rowCount ? paddingBottom : rowHeight}
+                            columnCount={columnCount}
+                            columnWidth={columnWidth}
+                            rowCount={rowCount}
+                            rowHeight={rowHeight}
+                            containerStyle={{
+                                width: columnCount * columnWidth + paddingRight,
+                                height: rowCount * rowHeight + paddingBottom,
+                                maxWidth: columnCount * columnWidth + paddingRight,
+                                maxHeight: rowCount * rowHeight + paddingBottom
+                            }}
                             onScroll={({ scrollTop, scrollLeft }) => onScroll({ scrollTop, scrollLeft })}
                             scrollTop={scrollTop}
                             scrollLeft={scrollLeft}
```

**Comment by bvaughn** [comment]
I'm afraid I put my foot in my mouth on this one. I actually cap the scroll offsets to make sure they don't exceed the bounds of the windowed collection. This means that the padded `Grid` won't dispatch a scroll event with offsets beyond the last cell. Normally this is desirable but in this case it leads to the behavior you've described. ☹️

I will give it some more thought. But it's looking like your original hack may still be the best way forward at the moment. My apologies.

**Comment by tlrobinson** [comment]
@bvaughn Thanks for looking into it.

BTW is a component like this (a grid with fixed column or row headers) in the scope of `react-virtualized`? Or perhaps just some props on Grid (`fixedRowCount`, `fixedColumnCount` or something) that create multiple Grids with ScrollSync internally? I'd be happy to help contribute it.

**Comment by bvaughn** [comment]
No problem @tlrobinson. Happy to help, even though I wasn't very helpful in this case. 😄 

I don't think adding `fixedRowCount` / `fixedColumnCount` props to `Grid` would work out very well. For one, people may want the fixed rows/columns to appear in different places than the top/left. But also it adds complexity to an already-very-complex component (`Grid`).

A decorator component that manages this is maybe more feasible.

**Pull request merged**